### PR TITLE
[Request Review] Distributed Counters and 2-Word Hashes

### PR DESCRIPTION
**Pull request by tfukaza**
**Edit (5/4/20)**
In retrospect this is pretty extra, if it seems like too much effort to merge I don't mind deleting the branch. 
---

This is a spin-off of the new-class branch, which implements a new mechanism to allocate WID to classes (and potentially programs). It has been updated to be compatible with the latest release branch. 

The problem with the current mechanism of randomly allocating WID is that:
-   The is a high chance of collision
-  ... which unfortunately seems to be what is happening, as for unknown reasons many WID's are being created an consecutive order (because if a collision happens, the default behavior is to check if the next WID is available)
- Requires 3 words to allocate a WID

The new mechanism addresses this issue by using a distributed counter to guarantee that a unique WID is generated every time. This allows about 16 million unique WIDs to be generated at roughly O(1) time, using only 2 words. In addition, a simple load balance mechanism ensures that WID's will be allocated evenly across the 16 million slots. The load balance also doubles as a shutoff mechanism, which will return an error if all slots are used up. Lastly, the counter calls the `tinycrypt` library to shuffle and WID and people who made requests simultaneously from getting similar WIDs. 

Also it should be fully backwards compatible too, ie all we need to do is add a single document called "--shard--" to the current database, and all database handlers will still work. 

Preliminary testing shows that with 8 shards, the system can take about 32 requests per second, which is about 6x the estimated maximum load on the system. 

The downside is the complexity, which is why this PR is [Request Review]. Also I am not sure atm if the read operation, which is done once per shard, will result in excessive read requests to Firebase

And obviously this is not really a high-priority feature, I just wanted to get it out for review before I lose track too much.

### CODE REVIEW

**Comment by krashanoff** [comment]
Thanks a ton Tomo! I'll be making my way through this PR slowly but surely. Should release be merged prior to this PR, I'll change the base and resolve any merge conflicts that may arise - though I suspect there won't be any.

File: alias_test.go [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+		for i := 0; i < 32; i++ {
+			u, err := d.GetID(context.Background(), "classes_alias")
+			if err != nil {
+				t.Fatal(err)
+			}
+			
+			t.Logf("u: %d\n",u)
+			w := tinycrypt.GenerateWord24(uint64(u))
+			wid := strings.Join(w, ",") 
+			t.Logf("u: %s\n======",wid)
```
Comment (krashanoff) [inline alias_test.go:49]: Making the test logs more descriptive would be pretty handy. On another note, having more tests in general would be handy, but we can add those retroactively seeing as top priority is getting the backend rolled out (finally).

File: db/aliasManager.go [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+	},
+}
+
+
+
+
+func (d *DB) InitShards(ctx context.Context, path string)(error){
+
+	// get the document that has the counters
+	doc := d.Collection(path).Doc("--shards--")
```
Comment (krashanoff) [inline db/aliasManager.go:34]: Might be worthwhile making the shard document name a constant.

File: db/aliasManager.go [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+
+	// get every document in the collection
+	shards := doc.Collection("shards").Documents(ctx)
+	count := make([]int64, num_shards)
+
+	// check each shard 
+	i := 0
+	for {
+
+		d, e := shards.Next()
```
Comment (krashanoff) [inline db/aliasManager.go:72]: Should probably put your incrementation at the bottom of the loop to ensure you reach all documents, but there may be a detail I'm missing here.

File: db/aliasManager.go [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+	// On success, a nil will be returned. 
+	err := d.RunTransaction(ctx, func(ctx context.Context, tx *firestore.Transaction) error{
+
+		doc, err := tx.Get(shard_ref)
+
+		if err != nil {
+			return err
+		}
+
+		tid, err := doc.DataAt("Count")
```
Comment (krashanoff) [inline db/aliasManager.go:138]: Should probably handle this potential error.

File: alias_test.go [hunk 5]
```diff
@@ -0,0 +1,55 @@
+package main_test
+
+import (
+	"context"
+	"testing"
+	"strings"
+
+	"./db"
```
Comment (krashanoff) [inline alias_test.go:8]: Since swapping to Go modules, we are using the `github.com/uclaacm/teach-la-go-backend/db` style of imports.

**Comment by krashanoff** [comment]
Just checking in on this PR: how are things going @tfukaza?

File: server_test.go [hunk 6]
```diff
@@ -150,6 +150,19 @@ func TestRigDB(t *testing.T) {
...
+
+		t.Log(rr.Body)
+
+		j, err := ioutil.ReadAll(rr.Result().Body)
+		if err != nil {
+			t.Fatal("Failed to read response")
+		}
+
+		json.Unmarshal([]byte(j), &class)
+		class_id = class.WID
```
[the remark this answers is not in the document]
Comment (tfukaza) [reply server_test.go]: Removed

File: alias_test.go (hunk 1 above)
Comment (tfukaza) [reply alias_test.go]: Updated the print function to be more descriptive about what the outputs are

File: db/aliasManager.go (hunk 2 above)
Comment (tfukaza) [reply db/aliasManager.go]: new constant `shard_name` created

File: db/aliasManager.go (hunk 3 above)
Comment (tfukaza) [reply db/aliasManager.go:71]: This was how they did it in the official tutorial, tbh I'm not confident enough to challenge professionals at Firebase. https://firebase.google.com/docs/firestore/solutions/counters

File: db/aliasManager.go (hunk 4 above)
Comment (tfukaza) [reply db/aliasManager.go:145]: Handled

File: alias_test.go (hunk 5 above)
Comment (tfukaza) [reply alias_test.go]: Updated to import from GitHub

**Comment by tfukaza** [comment]
One last worry I have is the `tinycrypt`. *Technically* people can look at the github and reverse-engineer the WID from the CID, which I assumed was not an issue since the classes don't really contain any confidential data, and is mostly public-facing. Correct me if I'm wrong.

File: db/aliasManager.go (hunk 3 above)
Comment (krashanoff) [reply db/aliasManager.go:71]: Probably all clear then. Actually makes sense on second thought now that I see that `d` and `e` are both initialized here.

**Pull request merged**